tsl/src/gapfill/planner.c

```diff
@@ -336,8 +425,82 @@ plan_add_gapfill(PlannerInfo *root, RelOptInfo *group_rel)
 
 		foreach (lc, copy)
 		{
-			add_path(group_rel, gapfill_path_create(root, lfirst(lc), context.call));
+			add_path(group_rel, gapfill_path_create(root, lfirst(lc), context.call.func));
 		}
 		list_free(copy);
 	}
 }
+
+static inline bool
+is_gapfill_path(Path *path)
+{
+	return IsA(path, CustomPath) && castNode(CustomPath, path)->methods == &gapfill_path_methods;
+}
+
+/*
+ * Since we construct the targetlist for the gapfill node from the
+ * final targetlist we need to adjust any intermediate targetlists
+ * between toplevel window agg node and gapfill node
+ */
+void
+gapfill_adjust_window_targetlist(PlannerInfo *root, RelOptInfo *input_rel, RelOptInfo *output_rel)
+{
+	ListCell *lc;
+
+	if (!is_gapfill_path(linitial(input_rel->pathlist)))
+		return;
+
+	foreach (lc, output_rel->pathlist)
+	{
+		WindowAggPath *toppath = lfirst(lc);
+
+		/*
+		 * if we have multiple window agg level we need to adjust targetlists of non-toplevel
+		 * WindowAggPaths
+		 */
+		if (IsA(toppath, WindowAggPath) && toppath->winclause->winref > 1)
+		{
+			WindowAggPath *path;
+			/*
+			 * we skip toplevel WindowAggPath because targetlist of toplevel WindowAggPath
```

Comment:
I would merge this comment with the one above the if

Reply:
done

Reply:
done